treeherder/auth/backends.py

```diff
@@ -37,48 +36,50 @@ class TaskclusterAuthBackend(object):
          'expires': '2016-10-31T17:40:45.692Z'}
     """
 
-    def _get_scope_value(self, scopes, scope_prefix):
-        for scope in scopes:
-            if scope.startswith(scope_prefix):
-                return scope[len(scope_prefix):]
-        return None
-
-    def _get_email(self, result):
+    def _extract_email_from_clientid(self, client_id):
         """
-        Get the user's email from the mozilla-user scope
+        Extract the user's email from the client_id
 
-        For more info on scopes:
-        https://docs.taskcluster.net/manual/3rdparty#authenticating-with-scopes
+        The client_id MUST be in the form:
+            "email/<username>@<domain>"
+              or
+            "mozilla-ldap/<username>@<domain>"
         """
-        email = self._get_scope_value(result["scopes"], "assume:mozilla-user:")
-
-        if email and re.search(r'.+@.+', email):
-            return email
-        raise TaskclusterAuthenticationFailed(
-            "Invalid email for clientId: '{}'. Invalid value for scope 'assume:mozilla-user': '{}'".format(
-                result["clientId"], email))
 
-    def _get_user(self, email):
+        if client_id.startswith("email/") or client_id.startswith("mozilla-ldap/"):
+            email = client_id.split("/", 1)[1]
+            # ensure what we find after the split is a valid email and ONLY
+            # a valid email
+            match = re.search(
+                r"(^[a-zA-Z0-9_.+-]+@[a-zA-Z0-9-]+\.[a-zA-Z0-9-.]+$)",
+                email)
+            if match:
+                return match.group(0)
+        raise NoEmailException("No email found in clientId: '{}'".format(client_id))
+
+    def _find_user_by_email(self, email, username, scopes):
         """
-        Try to find an exising user that matches the email.
-
-        TODO: Switch to using ``username`` instead of email once we are on
-        Django 1.10 in Bug 1311967.  will need to migrate existing hashed
-        usenames at that point.
-
+        Try to find an existing user that matches the email.
         """
 
-        # Since there is a unique index on username, but not on email,
-        # it is POSSIBLE there could be two users with the same email and
-        # different usernames.  Not very likely, but this is safer.
-        user = User.objects.filter(email=email)
+        if scope_match(scopes, [["assume:mozilla-user:{}".format(email)]]):
+            # Find the user by their email.
+
+            # Since there is a unique index on username, but not on email,
+            # it is POSSIBLE there could be two users with the same email and
+            # different usernames.  Not very likely, but this is safer.
+            users = User.objects.filter(email=email)
```

Comment:
Out of curiosity I checked to see if there were any duplicate emails on prod using:
```sql
SELECT username, email, COUNT(*) c FROM treeherder.auth_user GROUP BY email HAVING c > 1;
```
...which returned zero rows :-)

Even if there were duplicates, I wonder if we should error out rather than just picking the first email?

If it were desirable to error out, then the `.filter()` could be replaced by `.get()`, saving us needing to use `.first()` below, and we also don't then need to explicitly raise `ObjectDoesNotExist` since `.get()` does that for us.

Anyway, seeing as this code is only to help with the transition and will be removed after, it doesn't matter too much so happy with either :-)